### PR TITLE
implement spatial indexes

### PR DESCRIPTION
**Pull request by JCOR11599**
Changes:
 - Add feature flag to enabled spatial indexes: `DOLT_ENABLE_SPATIAL_INDEX`
 - Changes to flatbuffer stuff to read/write spatial indexes
 - First pass at naive and unoptimized implementation of Spatial Indexes
 - Tests for Spatial Indexes, Spatial Index Plans, and their Prepared counterparts (some are skipped)
 - Contains some fixes for typeinfo not retaining SRID information for multi-geometries
 - Changes to fix some diff table prepared issues

Companion PR: https://github.com/dolthub/go-mysql-server/pull/1558
Note: prepared test that I changed here had a column anmed `commit`, which got confused with `to_commit` in the diff https://github.com/dolthub/dolt/issues/5188

TODOs
 - Proper benchmarks
 - Potential Optimizations:
   - Level Pruning
   - Z-BBox skipping
   - Subdividing queries
   - Make Spatial Indexes Covering
   - Multi-Threading Level Searches

### CODE REVIEW

File: go/libraries/doltcore/schema/encoding/schema_marshaling.go [hunk 1]
```diff
@@ -151,6 +151,7 @@ type encodedIndex struct {
 	Tags            []uint64 `noms:"tags" json:"tags"`
 	Comment         string   `noms:"comment" json:"comment"`
 	Unique          bool     `noms:"unique" json:"unique"`
+	Spatial         bool     `noms:"spatial,omitempty" json:"spatial,omitempty"`
```
Comment (andy-wm-arthur) [inline go/libraries/doltcore/schema/encoding/schema_marshaling.go:154]: you should revert these changes, we don't support spatial indexes in the old format. you should also add some tests to confirm this

File: go/libraries/doltcore/schema/index_coll.go [hunk 2]
```diff
@@ -16,10 +16,20 @@ package schema
 
 import (
 	"fmt"
+	"os"
 	"sort"
 	"strings"
 )
 
+// EnableSpatialIndex enables the creation and use of spatial indexes
+var EnableSpatialIndex = false
```
Comment (andy-wm-arthur) [inline go/libraries/doltcore/schema/index_coll.go:25]: make a BATS test to confirm that this feature flag is working, both on and off

File: go/libraries/doltcore/schema/typeinfo/typeinfo.go [hunk 3]
```diff
@@ -181,13 +181,13 @@ func FromSqlType(sqlType sql.Type) (TypeInfo, error) {
...
-			return &multilinestringType{}, nil
+			return &multilinestringType{sqlType.(gmstypes.MultiLineStringType)}, nil
 		case gmstypes.MultiPolygonType{}.String():
-			return &multipolygonType{}, nil
+			return &multipolygonType{sqlType.(gmstypes.MultiPolygonType)}, nil
 		case gmstypes.GeomCollType{}.String():
-			return &geomcollType{}, nil
+			return &geomcollType{sqlType.(gmstypes.GeomCollType)}, nil
 		case gmstypes.GeometryType{}.String():
 			return &geometryType{sqlGeometryType: sqlType.(gmstypes.GeometryType)}, nil
```
Comment (andy-wm-arthur) [inline go/libraries/doltcore/schema/typeinfo/typeinfo.go:192]: what is this change for?

File: go/libraries/doltcore/sqle/enginetest/validation.go [hunk 4]
```diff
@@ -141,8 +141,10 @@ func validateIndexConsistency(
 	def schema.Index,
 	primary, secondary prolly.Map,
 ) error {
-	// TODO: fix this later
-	if len(def.PrefixLengths()) > 0 {
+	// TODO: the descriptors in the primary key are different
+	// than the ones in the secondary key; this test assumes
+	// they're the same
+	if len(def.PrefixLengths()) > 0 || def.IsSpatial() {
```
Comment (andy-wm-arthur) [inline go/libraries/doltcore/sqle/enginetest/validation.go:147]: implement the logic to validate both of these speacial case indexes, rather than just skipping them. otherwise you're eroding the value of this validation

File: go/libraries/doltcore/sqle/index/dolt_index.go [hunk 5]
```diff
@@ -910,14 +929,132 @@ func (di *doltIndex) trimRangeCutValue(to int, keyPart interface{}) interface{}
...
+func (di *doltIndex) prollySpatialRanges(ctx context.Context, ns tree.NodeStore, ranges []sql.Range, tb *val.TupleBuilder) ([]prolly.Range, error) {
+	// should be exactly one range
+	rng := ranges[0][0]
+	lower, upper := sql.GetRangeCutKey(rng.LowerBound), sql.GetRangeCutKey(rng.UpperBound)
+
+	// TODO: no clue what this is for
+	tup := tb.BuildPermissive(sharePool)
+	if err := PutField(ctx, ns, tb, 0, lower); err != nil {
 		return nil, err
 	}
```
Comment (andy-wm-arthur) [inline go/libraries/doltcore/sqle/index/dolt_index.go:941]: should be able to delete this...

File: go/libraries/doltcore/sqle/index/z_encoding.go [hunk 6]
```diff
@@ -154,28 +154,38 @@ func ZSort(points []types.Point) []types.Point {
 
 // ZCell converts the GeometryValue into a Cell
 // Note: there is an inefficiency here where small polygons may be placed into a level that's significantly larger
+// TODO: we could probably improve performance here by not using custom structs
```
Comment (andy-wm-arthur) [inline go/libraries/doltcore/sqle/index/z_encoding.go:157]: seems like you took care of this in `codec.go`?

File: go/libraries/doltcore/sqle/index/z_encoding.go [hunk 7]
```diff
@@ -154,28 +154,38 @@ func ZSort(points []types.Point) []types.Point {
...
 	zMin := ZValue(types.Point{X: bbox[0], Y: bbox[1]})
 	zMax := ZValue(types.Point{X: bbox[2], Y: bbox[3]})
 
+	// Level rounds up, and masking operates in pairs
 	cell := val.Cell{}
-	binary.BigEndian.PutUint64(cell.ZValue[:], zMin[0])
-	binary.BigEndian.PutUint64(cell.ZValue[8:], zMin[1])
 	if res := zMin[0] ^ zMax[0]; res != 0 {
-		cell.Level = byte(64 - bits.LeadingZeros64(res)/2)
+		lvl := (bits.Len64(res) + 1) >> 1
```
Comment (andy-wm-arthur) [inline go/libraries/doltcore/sqle/index/z_encoding.go:166]: `shamt`? could use some comments

File: go/libraries/doltcore/sqle/writer/prolly_index_writer.go [hunk 8]
```diff
@@ -247,6 +247,7 @@ type prollySecondaryIndexWriter struct {
 	name          string
 	mut           *prolly.MutableMap
 	unique        bool
+	spatial       bool
```
Comment (andy-wm-arthur) [inline go/libraries/doltcore/sqle/writer/prolly_index_writer.go:250]: is this field used? same for next file

File: go/store/val/codec.go [hunk 9]
```diff
@@ -626,29 +626,23 @@ func stringFromBytes(b []byte) string {
...
+// The lowest level (a point) is 0
+// The next 16 bytes is the z-value encoding of the minimum point of that subregion
+type Cell [cellSize]byte
 
 func compareCell(l, r Cell) int {
-	if l.Level != r.Level {
-		return int(l.Level - r.Level)
-	}
-	return bytes.Compare(l.ZValue[:], r.ZValue[:])
+	return bytes.Compare(l[:], r[:])
```
Comment (andy-wm-arthur) [inline go/store/val/codec.go:636]: I love to see this

File: go/libraries/doltcore/sqle/index/dolt_index.go [hunk 10]
```diff
@@ -910,14 +929,132 @@ func (di *doltIndex) trimRangeCutValue(to int, keyPart interface{}) interface{}
...
+	for lvl := 0; lvl < 32; lvl++ {
+		minVal, maxVal := make([]byte, 17), make([]byte, 17)
+		minVal[0], maxVal[0] = byte(lvl), byte(lvl)
+		shamt := lvl << 1
+		zMin[1] = (zMin[1] >> shamt) << shamt
+		zMax[1] = (zMax[1] >> shamt) << shamt
+		binary.BigEndian.PutUint64(minVal[1:], zMin[0])
+		binary.BigEndian.PutUint64(minVal[9:], zMin[1])
+		binary.BigEndian.PutUint64(maxVal[1:], zMax[0])
+		binary.BigEndian.PutUint64(maxVal[9:], zMax[1])
```
Comment (andy-wm-arthur) [inline go/libraries/doltcore/sqle/index/dolt_index.go:967]: the encoding logic here is obscure and difficult to parse for most readers. move this logic to helper functions in `z_encoding.go`

**Review by andy-wm-arthur — APPROVED** [review]
Generally looks good, some remaining work to clean this up, then ship it

File: go/libraries/doltcore/schema/typeinfo/typeinfo.go (hunk 3 above)
Comment (JCOR11599) [reply go/libraries/doltcore/schema/typeinfo/typeinfo.go:192]: This is something I should've implemented when I was originally implementing `MultiPoint`, `MultiLineString`, `MultiPolygon`, and `GeometryCollection`. When defining columns over these types, we're able to specify a default SRID, and `sqlType` is what remembers the columns user defined SRID.

Example: `create table t (p point srid 0)`

Having a SRID defined is required for Spatial Indexes to work.

**Review by andy-wm-arthur — APPROVED** [review]
